Allow lexical subroutines to be inside subroutines

nanto opened the pull request:
Lexical subroutines should be allowed to be inside other subroutines.

Fixes #946.

Fixed #972.

petdance commented:
What is the old behavior, without your patch?  Does it just not recognize the code?  Give an error?  Also, I think the test should check for more than just one nested sub, and more than one level of nested sub.

nanto commented:
Sorry for lack of context. Running the following code:

```perl
use Perl::Critic;
my $critic = Perl::Critic->new('-single-policy' => 'Subroutines::ProhibitNestedSubs');
my @violations = $critic->critique(\'sub foo { my sub bar { 1 } }');
print @violations;
```

Without this patch, this code shows a violation message `Nested named subroutine at line 1, column 11. Declaring a named sub inside another named sub does not prevent the inner sub from being global.`.  This message is wrong since lexical subroutines are not global,

With this patch, no message is shown.

nanto commented:
I added some cases to test. Please tell me if there are more cases we should test.

petdance commented:
I think we need to discuss implementation a bit more before we go ahead with this.  I'm continuing discussion back on #946.

petdance commented:
Cases we need to test (and you may covered some of these):

`sub A { my sub B {} }` = 0 errors

`sub A { sub B { } }` = 1 error

`sub A { sub B { sub C {} } }` = 2 errors on B and C

`sub A { my sub B {} }` = 0 errors

`sub A { my sub B { sub C {} } }` = 1 error on C

`sub A { sub B { my sub C {} } }` = 1 error on B

`sub A { my sub B { sub C {} } }` = 1 error on C

`sub A { sub B {} my sub C {} }` = 1 error on B

`sub A { my sub B {} sub C {} }` = 1 error on C

Any other combinations that I forgot?  I guess maybe we should check the cases where the top-level `sub A` is a lexical sub too.

nanto commented:
Thank you for listing test cases! I try to add them.

nanto commented:
Added test cases and rewrote implementation. 

I don't add the following case:

> `sub A { sub B { } }` = 1 error

since it is already covered in the "Basic failure" case.

nanto commented:
Thank you for merging!

petdance commented:
You're welcome. Thanks for contributing.
